### PR TITLE
[api-extractor] Add handling of file reference directives

### LINKED ISSUE #1761: [api-extract] reference to relative path is removed in roll up type definition
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [X] Bug

**Please describe the actual behavior.**

1. we have a source file `index.ts` that contains a reference to a relative path

https://github.com/jeremymeng/ae-repro/blob/80e5bfb2af8f9ab468f8ede8b763dc660945c369/src/index.ts

```
/// <reference path="../dom-shim.d.ts" />
```

2. the types generated by typescript compiler has the expected reference

https://github.com/jeremymeng/ae-repro/blob/80e5bfb2af8f9ab468f8ede8b763dc660945c369/src/index.d.ts#L1

```
/// <reference path="../dom-shim.d.ts" />
```

3. we roll up the type definition using api-extractor

https://github.com/jeremymeng/ae-repro/blob/80e5bfb2af8f9ab468f8ede8b763dc660945c369/api-extractor.json

The reference line is gone from the roll up d.ts file.

https://github.com/jeremymeng/ae-repro/blob/80e5bfb2af8f9ab468f8ede8b763dc660945c369/dist/index.d.ts

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

Repro project at https://github.com/jeremymeng/ae-repro.  Just run `npm run build` after `npm i`.

**What is the expected behavior?**

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** api-extractor
* **Tool Version:** "@microsoft/api-extractor": "^7.7.8"
* **Node Version:** v12.14.1
  * **Is this a LTS version?** 
  * **Have you tested on a LTS version?** 
* **OS:** Ubuntu 18

### PR DESCRIPTION
**Pull request by willmtemple**
This is a first stab at supporting path-based references like:

```typescript
/// <reference path="path/to/xyz.d.ts" />
```

in API Extractor.

Would fix: #1761 (CC @jeremymeng)

We use these references as part of a shimming pattern in the Azure SDK for JavaScript, but we aren't able to use dtsRollup for packages that need them because d.ts rollup currently just doesn't handle them.

Because these references are relative to the SourceFile that they are collected from, the path that is emitted has to be corrected to be relative to the ultimate file. I've written the logic for this as internal functions next to the Collector and DtsRollupGenerator, but could relocate them as needed.

I have not tested this on Windows yet.

### CODE REVIEW

File: apps/api-extractor/src/collector/Collector.ts [hunk 1]
```diff
@@ -43,6 +43,29 @@ export interface ICollectorOptions {
   extractorConfig: ExtractorConfig;
 }
 
+/**
+ * Resolve the name of a file to an absolute path relative to a TypeScript SourceFile node.
+ */
+function resolveNameRelativeToSourceFile(name: string, source: ts.SourceFile): string {
```
Comment (octogonz) [inline apps/api-extractor/src/collector/Collector.ts:49]: Can you simply use [path.relative()](https://nodejs.org/api/path.html#path_path_relative_from_to)?

File: apps/api-extractor/src/generators/DtsRollupGenerator.ts [hunk 2]
```diff
@@ -45,6 +45,35 @@ export enum DtsRollupKind {
   PublicRelease
 }
 
+/**
+ * Compute a relative path between two absolute paths.
+ */
+function computeRelativePath(fromPath: string, toPath: string): string {
```
Comment (octogonz) [inline apps/api-extractor/src/generators/DtsRollupGenerator.ts:51]: Can you simply use [path.relative()](https://nodejs.org/api/path.html#path_path_relative_from_to)?

**Comment by octogonz** [comment]
Thanks for making a PR! 😁

Could you add a test case to the **rushstack\build-tests\api-extractor-scenarios** folder?  (The test cases get registered in [rushstack\build-tests\api-extractor-scenarios\config\build-config.json](https://github.com/microsoft/rushstack/blob/a7546938f9745291b455546a3e15bb94661c727c/build-tests/api-extractor-scenarios/config/build-config.json).)

File: apps/api-extractor/src/collector/Collector.ts [hunk 3]
```diff
@@ -158,6 +182,17 @@ export class Collector {
...
   }
 
+  /**
+   * A list of names (e.g. "runtime-library") that should appear in a path-based reference like this:
+   *
+   * ```
+   * /// <reference path="runtime-library" />
+   * ```
+   */
+  public get dtsFileReferenceDirectives(): ReadonlySet<string> {
```
Comment (octogonz) [inline apps/api-extractor/src/collector/Collector.ts:192]: The `runtime-library` example is not intended to be a path.  Maybe something like this:
```ts
  /**
   * A list of paths (e.g. "path/to/file.d.ts") that should appear in a reference like this:
   *
   * ```
   * /// <reference path="path/to/file.d.ts" />
   * ```
   *
   * The paths are resolved relative to the source file containing the reference.
   */
```

File: apps/api-extractor/src/collector/Collector.ts (hunk 1 above)
Comment (willmtemple) [reply apps/api-extractor/src/collector/Collector.ts:49]: For this one, I don't think so. I need to boot into Windows and check it, but I believe that the reference directive in text has to follow the POSIX convention and use forward slash as its path separator just like a node module, but the file name of the source file could be in Windows convention, so relying on `path` to process the reference path may not be safe. Need to do some testing.

File: apps/api-extractor/src/generators/DtsRollupGenerator.ts (hunk 2 above)
Comment (willmtemple) [reply apps/api-extractor/src/generators/DtsRollupGenerator.ts:51]: In this instance, sure. The `Path` module in the node-core-library package in this repo made me think that there was some reason for avoiding it.

File: apps/api-extractor/src/collector/Collector.ts (hunk 3 above)
Comment (willmtemple) [reply apps/api-extractor/src/collector/Collector.ts:192]: ah yeah just missed this when copying the doc string :+1:

**Comment by willmtemple** [comment]
@octogonz I am getting a build failure on Windows in `node-core-library`. blocking `api-extractor-scenarios` from building.

I'm just building using `rush update` then `rush build` in Windows PowerShell, and I've set my execution policy to "Bypass" for the process. I don't know if there are other special instructions for building on windows because I can't find documentation on the build process.

Here is the error:

```
 FAIL  lib\test\Executable.test.js
  ● Executable.spawnSync("npm-binary-wrapper") edge cases 1

    Error

      86 |     expect(result.error).toBeUndefined();
      87 |     expect(result.stderr).toBeDefined();
    > 88 |     expect(result.stderr.toString()).toEqual('');
      89 |     expect(result.stdout).toBeDefined();
      90 |     const outputLines = result.stdout
      91 |         .toString()
      Error: expect(received).toEqual(expected)

      Expected value to equal:
        ""
      Received:
        "The directory name is invalid.
      "
      at executeNpmBinaryWrapper (lib/test/Executable.test.js:88:38)
      at Object.<anonymous> (lib/test/Executable.test.js:119:12)

  ● Executable.spawnSync("npm-binary-wrapper") edge cases 2

    Error

      86 |     expect(result.error).toBeUndefined();
      87 |     expect(result.stderr).toBeDefined();
    > 88 |     expect(result.stderr.toString()).toEqual('');
      89 |     expect(result.stdout).toBeDefined();
      90 |     const outputLines = result.stdout
      91 |         .toString()
      Error: expect(received).toEqual(expected)

      Expected value to equal:
        ""
      Received:
        "The system cannot find the path specified.
      "
      at executeNpmBinaryWrapper (lib/test/Executable.test.js:88:38)
      at Object.<anonymous> (lib/test/Executable.test.js:129:12)

  ● Executable.spawnSync("npm-binary-wrapper") edge cases 2

    Error

      86 |     expect(result.error).toBeUndefined();
      87 |     expect(result.stderr).toBeDefined();
    > 88 |     expect(result.stderr.toString()).toEqual('');
      89 |     expect(result.stdout).toBeDefined();
      90 |     const outputLines = result.stdout
      91 |         .toString()
      Error: expect(received).toEqual(expected)

      Expected value to equal:
        ""
      Received:
        "The system cannot find the path specified.
      "
      at executeNpmBinaryWrapper (lib/test/Executable.test.js:88:38)
      at Object.<anonymous> (lib/test/Executable.test.js:139:12)

 PASS  lib\test\PackageName.test.js
 PASS  lib\test\Text.test.js
 PASS  lib\test\ProtectableMap.test.js
 PASS  lib\test\PackageJsonLookup.test.js
 PASS  lib\test\Path.test.js
 PASS  lib\test\Sort.test.js
 PASS  lib\test\JsonSchema.test.js
 PASS  lib\test\FileSystem.test.js
[13:20:24] Error - 'jest' sub task errored after 8.75 s
 Jest tests failed
[13:20:24] 'test' errored after 18 s
[13:20:24]
About to exit with code: 1
Process terminated before summary could be written, possible error in async code not continuing!
Trying to exit with exit code 1
The script failed with exit code 1
```

File: apps/api-extractor/src/collector/Collector.ts (hunk 1 above)
Comment (octogonz) [reply apps/api-extractor/src/collector/Collector.ts:51]: If the input and output paths are POSIX format, then you can use `path.posix.relative()`.

If the input path is Windows style and you want to convert to POSIX, there is no theoretically correct API for that, but `Text.replaceAll(thePath, '\\', '/')` works reasonably well with Node.js

File: apps/api-extractor/src/generators/DtsRollupGenerator.ts (hunk 2 above)
Comment (octogonz) [reply apps/api-extractor/src/generators/DtsRollupGenerator.ts]: `FileSystem` is a complete replacement for `fs`, because the native `fs` library was designed as low-level primitives that are not suitable for everyday usage.

`Path` contains supplementary stuff that merely extends the native `path`, since that API is fine to use.

We should probably document this better.

**Comment by octogonz** [comment]
> I am getting a build failure on Windows in `node-core-library`. blocking `api-extractor-scenarios` from building.
> 
> I'm just building using `rush update` then `rush build` in Windows PowerShell, and I've set my execution policy to "Bypass" for the process. I don't know if there are other special instructions for building on windows because I can't find documentation on the build process.

@willmtemple I cannot repro this.  

What is the full path of your repo (`pwd` in PowerShell)?  I wonder if one of the parent paths has weird characters, or exceeds MAX_PATH.  I typically clone my repo as `C:\Git\rushstack` to minimize these issues.

**Comment by octogonz** [comment]
> I don't know if there are other special instructions for building on windows because I can't find documentation on the build process.

API Extractor docs are [here](https://api-extractor.com/pages/contributing/building/).  General repo instructions are [here](https://rushstack.io/pages/contributing/get_started/).

**Comment by octogonz** [comment]
@willmtemple Following up -- are you still working on this PR? Do you need anything else from us?

**Comment by willmtemple** [comment]
@octogonz Had to put this down for a while. The build issue I was running into was indeed a MAX_PATH issue. This week I'm going to revisit this. Referring to your comment in the linked issue:

> For a rollup, there's some question as to how those paths should be handled.

Do you have any thoughts on the two approaches to take here? This implementation is leaving the file references intact in the output bundle, just rewriting them to be relative to the output location, but the other alternative of course is that we could include the referenced file in the rolled-up d.ts.

**Comment by pd4d10** [comment]
This PR also seems to fix #3000 

Is there anything I can do to help?
